src/engine/JoinTree.hpp

```diff
@@ -0,0 +1,510 @@
+// Copyright 2024, University of Freiburg,
+// Chair of Algorithms and Data Structures.
+// Author:
+// Mahmoud Khalaf (2024-, khalaf@cs.uni-freiburg.de)
+
+#pragma once
+
+#include <algorithm>
+#include <iterator>
+#include <map>
+#include <numeric>
+#include <ranges>
+#include <set>
+#include <span>
+#include <vector>
+
+namespace JoinOrdering {
+
+enum class Direction {
+  UNDIRECTED,
+  PARENT,
+  CHILD,
+
+};
+
```

Comment:
Hi,
Some initial comments on the code design:

* please split this up into a .h and a .cpp file (or alternatively, if templating this into a similar *.h and *Impl.h file

* We should (for understandability) split the parts of the graph management, and the actual values. 
  So you need  a generic graph that has a generic payload for the nodes (relation names, cardinalities) as well as for the edges (here we will have the selectivities for the IKKBZ).

*

Reply:
* and for your precedence graph building, you should build a new tree, the in-place mutation is not easy to understand.

* so in general, it would be best to use a generic Graph interface that doesn't know anything about IKKBZ and then build IKKBZ on top of this